### DISCUSSION TITLE
About fault tolerance and high availability

### DISCUSSION THREAD

**Discussion by zbalkan** [stalwartlabs/smtp-server#13] [General]
Hi,

I wonder what is the suggested design for clustering for FT and HA to use in production. Documentation mentions single server and docker installation. But I couldn't find the details on load balancing anywhere. Can you point me where I can get the details?

**Comment by mdecimus** [comment]
Hi,

At the moment the only supported distributed configuration is load balancing, you can have multiple Stalwart SMTP servers each one with it's own MX record and the incoming messages will be distributed amongst them.
For high availability you need Kubernets or similar but currently the message queue is not replicated by default, it needs to be mounted on a distributed file system. However, I do plan to add built-in support for high availability once the all-in-one mail server is released.

**Reply by zbalkan** [reply]
Hi, 

Thank you for the information. I plan to use it in production but in a relatively simple way: an SMTP relay server for hybrid cloud setup. I can live with one instance for a while accepting the risk. I may use the 2 MX records solution since the load is not too high. But it would be great if I start with a proper cluster.

Also, do you think using an external queue like RabbitMQ or Logstash would help for your shared queue problem?

**Reply by mdecimus** [reply]
> Thank you for the information. I plan to use it in production but in a relatively simple way: an SMTP relay server for hybrid cloud setup. I can live with one instance for a while accepting the risk. I may use the 2 MX records solution since the load is not too high. But it would be great if I start with a proper cluster.

What you could do is set up a shared file system and keep the queue there. If there is a server failure a new instance of the SMTP server could be started pointing to the same queue. And if you'd like to scale horizontally you can add multiple instances of the server and balance the load between them.

> Also, do you think using an external queue like RabbitMQ or Logstash would help for your shared queue problem?

It is not a problem, it's just not implemented yet :) RabbitMQ/Kakfa could work for the metadata but a shared blob storage such as MinIO/S3 will be also required to store the messages.
